src/nanoarrow/common/array.c

```diff
@@ -456,10 +470,26 @@ static void ArrowArrayFlushInternalPointers(struct ArrowArray* array) {
   struct ArrowArrayPrivateData* private_data =
       (struct ArrowArrayPrivateData*)array->private_data;
 
-  for (int64_t i = 0; i < NANOARROW_MAX_FIXED_BUFFERS; i++) {
+  const bool is_binary_view = private_data->storage_type == NANOARROW_TYPE_STRING_VIEW ||
+                              private_data->storage_type == NANOARROW_TYPE_BINARY_VIEW;
+  const int32_t nfixed_buf = is_binary_view ? 2 : NANOARROW_MAX_FIXED_BUFFERS;
+
+  for (int32_t i = 0; i < nfixed_buf; i++) {
     private_data->buffer_data[i] = ArrowArrayBuffer(array, i)->data;
   }
 
+  if (is_binary_view) {
+    const int32_t nvirt_buf = private_data->n_variadic_buffers;
+    private_data->buffer_data = (const void**)ArrowRealloc(
+        private_data->buffer_data, sizeof(void*) * (nfixed_buf + nvirt_buf + 1));
```

Comment:
The allocation for `private_data->buffer_data` happens up front in `ArrowArrayInitFromType` and only reserves enough for `NANOARROW_FIXED_SIZE_BUFFERS`; in the case of having variadic buffers, you wouldn't really know what extra space needs to be allocated until you know how many variadic buffers you are dealing with.

Maybe there is another place you think is more suitable for this reallocation? But I don't _think_ this can be done up front in the Init function